### ISSUE TITLE
JIT: Stack zeroed with rep stosd

### ISSUE THREAD

**Issue by Zhentar** [dotnet/coreclr#19076]
C# source in this gist: https://gist.github.com/Zhentar/4ffb0a5d597c4c1e788d6007f1602b21

According to vTune, 5% of my execution time is in my function's prologue. This was unexpected because it hadn't been in previous iterations (and my function body had unfortunately not improved at all).
Looking at the the disassembly, I see:
```asm
LineEnumerator.MoveNext()
	push    rdi
	push    rsi
	sub     rsp,48h
	mov     rsi,rcx
	lea     rdi,[rsp+28h]
	mov     ecx,8
	xor     eax,eax
	rep     stos dword ptr [rdi]
	mov     rcx,rsi
	mov     rax,0F1CD0434ED23h
	mov     qword ptr [rsp+40h],rax
```

The `rep stos dword` in there seems rather odd - at the very least, it should be a `rep stos qword` with half as many iterations (although I'm not sure it would be any faster on my Skylake). But also I don't think there's any x86 architecture for which a 32 byte `rep stos` is faster than a reasonable unrolled version and the unrolled version wouldn't even be particularly large. And some of the comments in the JIT code seem to suggest that `rep stos` shouldn't ever be getting emitted.

**Comment by jkotas** [comment]
Related https://github.com/dotnet/coreclr/issues/13827
cc @erozenfeld

**Comment by AndyAyersMS** [comment]
In #19103 there are size wins (and likely speed wins) in minopts because reducing ref counts seems to encourage straight line expansions instead of loopy code. So I'd agree the heuristics for prolog initialization seem off.

**Comment by tannergooding** [comment]
@AndyAyersMS, is there any consideration done for the `ERMSB` (Enhanced REP MOVSB and STOSB operation) support?

**Comment by AndyAyersMS** [comment]
I haven't looked at this; @erozenfeld may have.

**Comment by fiigii** [comment]
> The rep stos dword in there seems rather odd - at the very least, it should be a rep stos qword with half as many iterations (although I'm not sure it would be any faster on my Skylake).

@Zhentar `rep stos*/movs*` are implemented in micro-code that the data type does not matter on performance, and actually `rep stosb/movsb` is the best choice on newer machines, please see https://github.com/dotnet/coreclr/pull/19655

**Comment by AndyAyersMS** [comment]
Still worth considering, but we're unlikely to get to this in 3.0, so marking as future.

**Comment by Zhentar** [comment]
I'd been wanting to run some perf tests around this, but I didn't want to deal with getting good numbers on it without BenchmarkDotNet helping me out.... so I figured out a way to do inline assembly in C#: https://github.com/Zhentar/CLRStackZeroingAssemblyTests

"Overhead" is the cost of a `calli` pinvoke delegate calling a function that is just a single `retn` instruction. In other words, the first 7ns of each result is test overhead; differences are proportionally larger than they appear

"Baseline" is identical to the code in my initial post, and what is still generated today.

"ImprovedStosD" is the same as baseline except that it doesn't unconditionally copy RCX to a non-volatile register. StosB and StosQ variants are what you'd expect.

"SSE2" is however many `vmovdqu` in a row it takes to zero the desired bytes

Conclusions:
* B/D/Q doesn't matter at all on my skylake
* Even though I have ERMSB, rep mov needs 512 bytes just to catch up with unrolled SSE sets
* ERMSB has a high startup cost. For just 32 bytes as in my first post, it's ~20x slower than the unrolled SSE sets

``` ini
BenchmarkDotNet=v0.11.4, OS=Windows 10.0.17134.590 (1803/April2018Update/Redstone4)
Intel Core i7-6600U CPU 2.60GHz (Skylake), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=3.0.100-preview-010184
  [Host]     : .NET Core 3.0.0-preview-27324-5 (CoreCLR 4.6.27322.0, CoreFX 4.7.19.7311), 64bit RyuJIT
  Job-VOVHWP : .NET Core 3.0.0-preview-27324-5 (CoreCLR 4.6.27322.0, CoreFX 4.7.19.7311), 64bit RyuJIT

MaxRelativeError=0.05  EnvironmentVariables=COMPlus_TieredCompilation=0  

```
|        Method | Bytes |      Mean |     Error |    StdDev |
|-------------- |------------ |----------:|----------:|----------:|
|      Overhead |          -  |  6.951 ns | 0.3478 ns | 0.3083 ns |
|      Baseline |          32 | 15.932 ns | 0.6266 ns | 0.5861 ns |
| ImprovedStosB |          32 | 16.552 ns | 1.1279 ns | 1.0550 ns |
| ImprovedStosD |          32 | 18.884 ns | 2.0452 ns | 1.8130 ns |
| ImprovedStosQ |          32 | 17.220 ns | 1.1702 ns | 1.0946 ns |
|          SSE2 |          32 |  7.300 ns | 0.1628 ns | 0.1523 ns |
|      Baseline |          64 | 17.830 ns | 1.4129 ns | 1.3216 ns |
| ImprovedStosB |          64 | 14.890 ns | 0.4814 ns | 0.4503 ns |
| ImprovedStosD |          64 | 15.607 ns | 0.7108 ns | 0.6649 ns |
| ImprovedStosQ |          64 | 15.217 ns | 0.4941 ns | 0.4380 ns |
|          SSE2 |          64 |  8.847 ns | 0.7380 ns | 0.6904 ns |
|      Baseline |         128 | 15.200 ns | 0.2035 ns | 0.1804 ns |
| ImprovedStosB |         128 | 15.451 ns | 0.5030 ns | 0.4459 ns |
| ImprovedStosD |         128 | 15.138 ns | 0.5372 ns | 0.4486 ns |
| ImprovedStosQ |         128 | 16.514 ns | 1.4771 ns | 1.3817 ns |
|          SSE2 |         128 | 10.419 ns | 0.4886 ns | 0.4331 ns |
|      Baseline |         256 | 20.407 ns | 1.9412 ns | 1.7208 ns |
| ImprovedStosB |         256 | 20.712 ns | 0.8683 ns | 0.8122 ns |
| ImprovedStosD |         256 | 18.140 ns | 0.4299 ns | 0.4021 ns |
| ImprovedStosQ |         256 | 18.058 ns | 0.1706 ns | 0.1595 ns |
|          SSE2 |         256 | 13.189 ns | 0.1245 ns | 0.1104 ns |
|      Baseline |         512 | 20.021 ns | 0.3746 ns | 0.3321 ns |
| ImprovedStosB |         512 | 19.949 ns | 0.8585 ns | 0.7611 ns |
| ImprovedStosD |         512 | 20.589 ns | 1.5961 ns | 1.4930 ns |
| ImprovedStosQ |         512 | 20.577 ns | 0.9243 ns | 0.8646 ns |
|          SSE2 |         512 | 20.348 ns | 1.3204 ns | 1.1705 ns |
|      Baseline |        1024 | 28.049 ns | 0.2983 ns | 0.2791 ns |
| ImprovedStosB |        1024 | 27.205 ns | 0.3070 ns | 0.2872 ns |
| ImprovedStosD |        1024 | 26.780 ns | 0.1957 ns | 0.1831 ns |
| ImprovedStosQ |        1024 | 22.553 ns | 0.2013 ns | 0.1883 ns |
|          SSE2 |        1024 | 34.732 ns | 2.1479 ns | 2.0091 ns |

**Comment by fiigii** [comment]
> ERMSB has a high startup cost. For just 32 bytes as in my first post, it's ~20x slower than the unrolled SSE sets

Right, ERMSx are not good on short strings, which was designed for long strings but Intel is continuously optimizing them(B).

**Comment by pentp** [comment]
From these stats it looks like `rep stos` in the prologue is quite bad for frame sizes below 512, but what's also interesting is the exponential slowdown for unrolled SSE2 (very noticeable at just 16 ops already) which might indicate a decoder exhaustion or some other interaction with the benchmark calling code?
In real world code this slowdown might bite a lot earlier, so I think this is one of those cases where a micro-benchmark might mislead.
A while ago when experimenting with RyuJIT I ran a decimal benchmark using both `rep stosb` and `rep stosq` and didn't notice much difference (`stosq` was a bit slower even IIRC), but such an approach could be extended to many different variants easily and use ASP.NET plaintext benchmark for example.

**Comment by Zhentar** [comment]
It is indeed interesting - I believe what's happening there is that there is effectively a threshold below which pipelining can hide the entire latency of the operations within the benchmark overhead instructions. I later managed to shave off a couple more overhead instructions, which gets the effect to reveal itself at 8 ops instead of 16:

|          Case | Size |      Mean |     Error |    StdDev |
|-------------- |----- |----------:|----------:|----------:|
|      Overhead |      |  1.781 ns | 0.1919 ns | 0.3309 ns |
|          SSE2 |   32 |  1.770 ns | 0.1706 ns | 0.2501 ns |
|          SSE2 |   64 |  2.453 ns | 0.2152 ns | 0.4857 ns |
|          SSE2 |  128 |  3.113 ns | 0.2420 ns | 0.3313 ns |
